qiskit/algorithms/minimum_eigensolvers/vqe.py

```diff
@@ -0,0 +1,380 @@
+from qiskit.utils import algorithm_globals
+
+from ..exceptions import AlgorithmError
+from ..list_or_dict import ListOrDict
+from ..optimizers import Optimizer, Minimizer, SLSQP
+from ..variational_algorithm import VariationalAlgorithm, VariationalResult
+from .minimum_eigensolver import MinimumEigensolver, MinimumEigensolverResult
+
+# from qiskit.algorithms.observables_evaluator import eval_observables
+
+logger = logging.getLogger(__name__)
+
+
+class VQE(VariationalAlgorithm, MinimumEigensolver):
+    r"""The variational quantum eigensolver (VQE) algorithm.
+
+    VQE is a quantum algorithm that uses a variational technique to find the minimum eigenvalue of
+    the Hamiltonian :math:`H` of a given system [1].
+
+    An instance of VQE requires defining two algorithmic sub-components: a trial state (a.k.a.
+    ansatz) which is a :class:`QuantumCircuit`, and one of the classical
+    :mod:`~qiskit.algorithms.optimizers`.
+
+    The ansatz is varied, via its set of parameters, by the optimizer, such that it works towards a
+    state, as determined by the parameters applied to the ansatz, that will result in the minimum
+    expectation value being measured of the input operator (Hamiltonian).
+
+    The optimizer can either be one of Qiskit's optimizers, such as
+    :class:`~qiskit.algorithms.optimizers.SPSA` or a callable with the following signature:
+
+    .. note::
+
+        The callable _must_ have the argument names ``fun, x0, jac, bounds`` as indicated
+        in the following code block.
+
+    .. code-block:: python
+
+        from qiskit.algorithms.optimizers import OptimizerResult
+
+        def my_minimizer(fun, x0, jac=None, bounds=None) -> OptimizerResult:
+            # Note that the callable *must* have these argument names!
+            # Args:
+            #     fun (callable): the function to minimize
+            #     x0 (np.ndarray): the initial point for the optimization
+            #     jac (callable, optional): the gradient of the objective function
+            #     bounds (list, optional): a list of tuples specifying the parameter bounds
+
+            result = OptimizerResult()
+            result.x = # optimal parameters
+            result.fun = # optimal function value
+            return result
+
+    The above signature also allows one to directly pass any SciPy minimizer, for instance as
+
+    .. code-block:: python
+
+        from functools import partial
+        from scipy.optimize import minimize
+
+        optimizer = partial(minimize, method="L-BFGS-B")
+
+    Attributes:
+        estimator: The estimator primitive to compute the expectation value of the circuits.
+        ansatz: The parameterized circuit used as an ansatz for the wave function.
+        optimizer: A classical optimizer to find the minimum energy. This can either be a
+            Qiskit :class:`.Optimizer` or a callable implementing the :class:`.Minimizer` protocol.
```

Comment:
Should this also document the `SLSQP` default?

Reply:
I'm actually going to remove this as a default, and force keywords after the optimizer. I.e.,

```python
 def __init__(
        self,
        estimator: BaseEstimator,
        ansatz: QuantumCircuit | None,
        optimizer: Optimizer | Minimizer,
        *
        gradient: BaseEstimatorGradient | None = None,
        initial_point: Sequence[float] | None = None,
        callback: Callable[[int, np.ndarray, float, float], None] | None = None,
```

The ansatz will be allowed to be `None` for the Nature factories. I've already discussed this with @woodsp-ibm, but feel free to comment on this approach, @mrossinek.